### PR TITLE
Esp32-S2 I2C fixes

### LINKED ISSUE #3846: Crash when repeatedly creating and destroying busio.I2C object
Running the following sequence twice in the same REPL causes CircuitPython to reset (not into safe mode, just a "normal" reset!)

Tested on a kaluga with `DEBUG=1` at f2204d7d8

(tio is my terminal program)
```
>>> import busio, board; bus = busio.I2C(board.IO10, board.IO11); bus.try_lock(); bus.scan(); bus.deinit()
True
[11, 112]
>>> import busio, board; bus = busio.I2C(board.IO10, board.IO11); bus.try_lock(); bus.scan(); bus.deinit()
True

[tio 10:05:07] Disconnected
[tio 10:05:41] Connected
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.

Press any key to enter the REPL. Use CTRL-D to reload.
```

gdb debugging doesn't show anything interesting.  hardware debug console doesn't show anything interesting.  I don't know how to continue debugging this sort of reset (e.g., to set a breakpoint to "catch" it)

```
I (15316) gpio: GPIO[11]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (15316) gpio: GPIO[10]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (20846) gpio: GPIO[11]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (20846) gpio: GPIO[10]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
<----- ( second bus.scan() call happens here) ---->
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x8 (TG1WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40028200
SPIWP:0xee
mode:DIO, clock div:2
load:0x3ffe6100,len:0x8
load:0x3ffe6108,len:0x1838
load:0x4004c000,len:0x944
load:0x40050000,len:0x2cec
entry 0x4004c1bc
I (51) boot: ESP-IDF v4.3-dev-1198-gd06744f5e 2nd stage bootloader
```

### LINKED ISSUE #4046: ESP32S2: I2C bus object crashes or failes when used after soft reboot
This code works fine when the MagTag boots cold.

```
import board
import busio
print(dir(board))
i2c = busio.I2C(board.SCL, board.SDA)

def printdevices():
    i2c.try_lock()
    # Find the I2C devices available.
    devices = i2c.scan()
    print("I2C devices:", len(devices))
    for device in devices:
        print(hex(device))
    i2c.unlock()
    print("I2C unlocked! ")

printdevices()
```

image

After connecting to Putty (here with COM29) 
and pressing ctrl-c to get to the CP prompt
and pressing ctrl-d to reload...

There is a core code hard crash:
image

And the CDC connection to Putty is disconnecting:
image

PS: I observe similar behaviour using a featherS2.

### PR DESCRIPTION
**Pull request by dhalbert**
Fixes #3846.
Fixes #4046.

- Reverts I2C driver management to use `i2c_driver_install()` and `i2c_driver_delete()`.
- Makes a change in `i2c.c` in the ESP-IDF about how a semaphore is deleted.
- Uses a mutex semaphore for the I2C lock, as recommended by @igrr in https://github.com/adafruit/circuitpython/issues/3846#issuecomment-749143943.

Tagging @askpatrickw, @microdev1, @skieast for testing.

This is a draft because it depends on https://github.com/adafruit/esp-idf/pull/1, which is not yet done. This build temporarily uses https://github.com/dhalbert/esp-idf/tree/circuitpython-i2c-cmd_mux-fix. 

Tests that now work:
```python
import board,busio

i2c = busio.I2C(board.SCL, board.SDA)
i2c.deinit()

import wifi
```
```python
import time
import board
from busio import I2C

def scan(deinit=True):
    i2c = I2C(board.SCL, board.SDA)
    i2c.try_lock()
    results = i2c.scan()
    if deinit:
        i2c.deinit()
    return results

print(scan())
print(scan())
print(scan())
```
```python
import board,busio,adafruit_sht31d

import wifi

i2c = busio.I2C(board.SCL, board.SDA)
i2c.deinit()

i2c = busio.I2C(board.SCL, board.SDA)
s = adafruit_sht31d.SHT31D(i2c)
print(s.temperature)

i2c.deinit()
```

### CODE REVIEW

File: ports/esp32s2/common-hal/busio/I2C.c [hunk 1]
```diff
@@ -149,12 +142,14 @@ void common_hal_busio_i2c_deinit(busio_i2c_obj_t *self) {
         return;
     }
 
-    i2c_status[self->i2c_num] = STATUS_FREE;
+    i2c_driver_delete(self->i2c_num);
 
-    common_hal_reset_pin(self->sda_pin);
     common_hal_reset_pin(self->scl_pin);
+    common_hal_reset_pin(self->sda_pin);
```
Comment (jepler) [inline ports/esp32s2/common-hal/busio/I2C.c:148]: If this change of order is important, can you comment about it?

Comment (dhalbert) [reply ports/esp32s2/common-hal/busio/I2C.c:148]: It is not important - I can undo that to make it consistent with the ordering elsewhere.

**Comment by askpatrickw** [comment]
@dhalbert  I have some time and will test this now

**Review by askpatrickw — COMMENTED** [review]
I tested this using the CI artifacts on a MetroS2 an a Feather S2 PreRelease.
I ran dan's second code sample, and my own script using the BME_680. Both worked as expected.

I then copied over a script that connects to wifi and then prints the BME values.
It failed to safemode, when I disconnected USB and connected again the script worked.
I tried to reproduce that auto-load failure 10 times and could not.

Unless other people see that behavior, I would call this good.

**Comment by skieast** [comment]
tested on UM FeatherS2.
Script uses wifi, AdaFruit_requests, then setup two i2c devices on different busses. AHt20 and SGP30
Loop pinging google and reading both devices.  Works. Interrupt with CTL-C, restart wih CTRL-D, works
From the repl create two i2c busses, i2c and i2c2.   try_lock and unlock work on both. 
Looks great.

**Comment by dhalbert** [comment]
Thank you very much for the testing! I've now moved this up to esp-idf merge commit (code is identical), and turned off draft status.

**Comment by dhalbert** [comment]
When #4378 is changed and merged, it will change from using bitbangio for 0, 1, or 2-byte writes to just 0-byte writes. I will try to test the equivalent of that change with some of the failing peripherals above.

**Review by microDev1 — COMMENTED** [review]
Thanks! @dhalbert.
I tested all three provided examples on my microS2 and they work flawlessly. Just one question...

**Comment by microDev1** [comment]
@dhalbert I think we can get the idf patch in v4.3 since its a pre-release at the moment.

File: ports/esp32s2/common-hal/busio/I2C.c [hunk 2]
```diff
@@ -106,34 +104,29 @@ void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
...
+
+    // Delete any previous driver.
+    i2c_driver_delete(self->i2c_num);
+
     i2c_config_t i2c_conf = {
         .mode = I2C_MODE_MASTER,
         .sda_io_num = self->sda_pin->number,
         .scl_io_num = self->scl_pin->number,
-        .sda_pullup_en = GPIO_PULLUP_DISABLE,  /*!< Internal GPIO pull mode for I2C sda signal*/
-        .scl_pullup_en = GPIO_PULLUP_DISABLE,  /*!< Internal GPIO pull mode for I2C scl signal*/
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/esp32s2/common-hal/busio/I2C.c:114]: That was a mistake; I think I was copying over some code from MicroPython to test. I'll put it back. Thanks for catching that.

**Comment by dhalbert** [comment]
> @dhalbert I think we can get the idf patch in v4.3 since its a pre-release at the moment.

I agree, it would be good to PR upstream, but I am still somewhat bothered by the whole thing because I don't see _why_ the IDF fix works. I did some more experiments yesterday and re-read a bunch of our code and their code. I came to no new conclusion. I haven't gotten feedback from igrr about it. Also I inquired with FreeRTOS: https://forums.freertos.org/t/vsemaphoredelete-immediately-after-xsemaphoretake/11956.

Ideally I would like to reproduce this in a simple ESP-IDF program, because they might ask for a test case, but I don't have any experience writing one. I could submit a PR and see what happens, but the fix seems a bit superstitious. There are other examples of `Take` followed by `Delete` in the ESP-IDF, mostly in test code. There are also a couple of examples of `Take`-`Give`-`Delete`. Do you have an intuition about this?

**Review by microDev1 — APPROVED** [review]
Thanks! for making the changes.

**Comment by microDev1** [comment]
I found that they use only `Delete` in the `err` routine of `i2c_driver_install` ([here](https://github.com/espressif/esp-idf/blob/73db142403c6e5b763a0e1c07312200e9b622673/components/driver/i2c.c#L347-L372)) for clean-up.
Getting someone from espressif to comment on why `Take` is required in case of `i2c_driver_delete` might clear things up.

**Comment by microDev1** [comment]
> Do you have an intuition about this?

No clue at the moment, I'll try to replicate this behavior in an `IDF` program.

**Comment by dhalbert** [comment]
Failed build was a network problem. Merging.

**Comment by dhalbert** [comment]
> No clue at the moment, I'll try to replicate this behavior in an `IDF` program.

When I look at the multiple threads running in a "crashed" program vs one running this fix, they look the same. In other words, it is not a hard crash, but some kind of loop or other problem. The clue with my simple tests above is that the REPL stops working, and eventually disconnects. So perhaps some program that does some USB serial activitiy would hang up. Perhaps also some wifi or I2C activity would fail - that might be easier to test.

**Comment by dhalbert** [comment]
This is the simplest example that reliably fails without the fix. It prints `>>>` after the `import wifi` but then the REPL hangs.
```python
import board,busio

i2c = busio.I2C(board.SCL, board.SDA)
i2c.deinit()

import wifi
```

**Comment by microDev1** [comment]
> It prints >>> after the import wifi but then the REPL hangs.

Yes, I get the same behavior... it hangs and no backtrace. From the debug print I see that wifi starts initializing but stops abruptly.
This also happens on `V4.3`. I am going to replicate the sequence of code that is executed with these commands.

**Comment by dhalbert** [comment]
So the essence of the above is:
```c
// From busio.I2C().
i2c_driver_delete(NULL);   // does nothing
i2c_param_config();
i2c_driver_install();
// i2c.deinit().
i2c_driver_delete();

// Below from common_hal_wifi_init() and common_hal_wifi_radio_set_enabled().
esp_netif_init();               
esp_event_loop_create_default();
 xEventGroupCreateStatic();
esp_event_handler_instance_register();
esp_event_handler_instance_register();
esp_wifi_init();
esp_wifi_set_mode();
esp_wifi_start();

**Comment by microDev1** [comment]
> So the essence of the above is

Yup! that's it... you want to go ahead with the `IDF` program? If not I can run this sequence tomorrow.

**Comment by dhalbert** [comment]
I'll try to get the time; I'll let you know. I guess I can gut one of their examples and put that in.

**Comment by skieast** [comment]
Glad someone is also wondering why it works.  I just got in and all my questions have been posed.

**Comment by dhalbert** [comment]
Moving this discussion to a new issue, #4401.

**Pull request merged**